### PR TITLE
Phonebook registration

### PR DESCRIPTION
**Pull request by shahrukh-ijaz**

### CODE REVIEW

File: contacts/views.py [hunk 1]
```diff
@@ -168,3 +169,17 @@ def email_delete(request, id, contact_id):
         email.delete()
     return HttpResponseRedirect('/display_contact/' + contact_id)
 
+
+def signup(request):
+    if request.method == 'POST':
```
Comment (asamolion) [inline contacts/views.py:175]: As always, please use a decorator.

File: contacts/tests.py [hunk 2]
```diff
@@ -1,12 +1,53 @@
...
 
 
 class ContactTestCase(TestCase):
+
+    def setUp(self):
+        user = User(username='sharukh', password='shahrukh31')
+        user.save()
+        Contact.objects.create(first_name="shahrukh", last_name="ijaz", note="mine",  dob='2019-03-14', user_id=user)
+
+    def able_to_delete_contact(self):
```
Comment (asamolion) [inline contacts/tests.py]: Any method of any sub-class of `TestCase` which doesn't start with `test_` is not considered a test case. Please rename all such test cases as such.

File: contacts/tests.py [hunk 3]
```diff
@@ -1,12 +1,53 @@
...
+        self.assertFalse(form.is_valid())
+
+
+class EmailTestCase(TestCase):
     def setUp(self):
-        Contact.objects.create(first_name="shahrukh", last_name="ijaz", note="mine")
+        Email.objects.create(Email='shahrukh.ijaz@arbisoft.com', contact_id=1)
+
+    def ErrorEmail(self):
+        email = Email.objects.get(Email="shahrukh.ijaz@arbisoft.com")
```
Comment (asamolion) [inline contacts/tests.py]: Don't interchange double and single quotes. Use, a single scheme and be consistent.

**Review by asamolion — CHANGES_REQUESTED** [review]

File: contacts/views.py (hunk 1 above)
Comment (shahrukh-ijaz) [reply contacts/views.py:175]: done

File: contacts/tests.py (hunk 2 above)
Comment (shahrukh-ijaz) [reply contacts/tests.py]: done

File: contacts/tests.py (hunk 3 above)
Comment (shahrukh-ijaz) [reply contacts/tests.py]: ok done

File: contacts/tests.py [hunk 4]
```diff
@@ -1,12 +1,55 @@
...
+        self.assertFalse(email, 'The email not belong to any Contact')
+
+    def test_successEmail(self):
+        email = Email.objects.create(Email='shahrukh.ijaz@arbisoft.com', contact_id=1)
+        self.assertEqual(email, 'Email Created!')
+
+
+class NumberTestCase(TestCase):
+
+    def test_setUp(self):
```
Comment (asamolion) [inline contacts/tests.py]: The setUp and tearDown methods are not prefixed with `test_`

**Review by asamolion — CHANGES_REQUESTED** [review]

**Review by asamolion — APPROVED** [review]

File: contacts/tests.py [hunk 5]
```diff
@@ -1,12 +1,52 @@
...
+    def test_to_delete_contact(self):
+        users = Contact.objects.filter(user_id=2).count()
+        if users > 0:
+            self.assertFalse(users, 'The contact have contacts able to delete!')
+        else:
+            self.assertFalse(users, 'The contact have contacts not able to delete!')
+
+    def test_invalid_form(self):
+        user = User.objects.create(username='sharukh', password='shahrukh31')
+        w = Contact.objects.create(first_name='shahrukh', last_name='ijaz', note='mine', dob='2019-03-14', user_id=user)
```
Comment (asamolion) [inline contacts/tests.py]: you don't need the `w` variable.

File: contacts/tests.py [hunk 6]
```diff
@@ -1,12 +1,52 @@
...
+
+
+class EmailTestCase(TestCase):
+
     def setUp(self):
-        Contact.objects.create(first_name="shahrukh", last_name="ijaz", note="mine")
+        user = User.objects.create(username='sharrukh9990', password='shahrukh31')
+        contact = Contact.objects.create(first_name='shahrukh', last_name='ijaz', note='mine', dob='2019-03-14', user_id=user)
+
+    def test_successEmail(self):
```
Comment (asamolion) [inline contacts/tests.py]: always use snake case in python.

File: contacts/tests.py [hunk 7]
```diff
@@ -1,12 +1,52 @@
...
+    def test_invalid_form(self):
+        user = User.objects.create(username='sharukh', password='shahrukh31')
+        w = Contact.objects.create(first_name='shahrukh', last_name='ijaz', note='mine', dob='2019-03-14', user_id=user)
+        form = AddContactForm()
+        self.assertFalse(form.is_valid())
+
+
+class EmailTestCase(TestCase):
+
     def setUp(self):
```
Comment (asamolion) [inline contacts/tests.py]: The setUp method is unnecessary in this test case.

File: contacts/tests.py [hunk 8]
```diff
@@ -1,12 +1,52 @@
...
+    def test_successEmail(self):
+        user = User.objects.create(username='sharuukh9990', password='shahrukh31')
+        contact = Contact.objects.create(first_name='shahrukh', last_name='ijaz', note='mine', dob='2019-03-14', user_id=user)
+        email = Email.objects.create(email='shahrukh.ijaz@arbisoft.com', contact_id=contact)
+        self.assertEqual('Email Created!', 'Email Created!')
+
+
+class NumberTestCase(TestCase):
+
+    def setUp(self):
```
Comment (asamolion) [inline contacts/tests.py]: The setUp method is unnecessary here as well.

File: contacts/tests.py [hunk 9]
```diff
@@ -1,12 +1,52 @@
...
+    def test_successNumber(self):
 
-    def test_contact_create(self):
-        """contact could be created are correctly identified"""
-        first_name = Contact.objects.get(first_name="shahrukh")
-        self.assertEqual(first_name.test_contact_create(), 'The contact can be created!')
+        user = User.objects.create(username='sharukhuu9990', password='shahrukh31')
+        contact = Contact.objects.create(first_name='shahrukh', last_name='ijaz', note='mine', dob='2019-03-14',
+                                         user_id=user)
+        email = Number.objects.create(number='03209503962', contact_id=contact)
```
Comment (asamolion) [inline contacts/tests.py]: The `email` variable is unnecessary.

**Review by asamolion — CHANGES_REQUESTED** [review]

File: contacts/tests.py (hunk 5 above)
Comment (shahrukh-ijaz) [reply contacts/tests.py]: done

File: contacts/tests.py (hunk 6 above)
Comment (shahrukh-ijaz) [reply contacts/tests.py]: done

File: contacts/tests.py (hunk 7 above)
Comment (shahrukh-ijaz) [reply contacts/tests.py]: ok

File: contacts/tests.py (hunk 8 above)
Comment (shahrukh-ijaz) [reply contacts/tests.py]: ok

File: contacts/tests.py (hunk 9 above)
Comment (shahrukh-ijaz) [reply contacts/tests.py]: ok

**Review by asamolion — APPROVED** [review]

**Pull request merged**